server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java

```diff
@@ -1598,4 +1599,18 @@ protected <Request extends ActionRequest, Response extends ActionResponse> void
             }
         };
     }
+
+    /**
+     * Same as {@link PlainActionFuture} but for use with {@link RefCounted} result types. Unlike {@code PlainActionFuture} this future
+     * acquires a reference to its result. This means that the result reference must be released by a call to {@link RefCounted#decRef()}
+     * on the result before it goes out of scope.
+     * @param <R> reference counted result type
+     */
+    private static class RefCountedFuture<R extends RefCounted> extends PlainActionFuture<R> {
+        @Override
+        public final void onResponse(R result) {
+            result.incRef();
+            super.onResponse(result);
```

Comment:
Maybe this'd be a little safer?

```suggestion
            assert result.hasReferences();
            if (set(result)) {
                result.incRef();
            }
```